[Docs][Kuberay] Documentation for Using Kuberay with TPUs (#45924)
ray-project/ray: doc/source/cluster/kubernetes/user-guides/gcp-gke-tpu-cluster.md

(kuberay-gke-tpu-cluster-setup)=

> # Start Google Cloud GKE Cluster with TPUs for KubeRay

Reviewer: It would be awesome if this guide show-cased a real example of using TPUs beyond just creating a GKE cluster with KubeRay and  TPUs enabled

Author: Added a Stable Diffusion example in[4296124](https://github.com/ray-project/ray/pull/45924/commits/42961243c485652ae4d35de1aa3779f12b199aed).

GKE uses Kubernetes node selectors to ensure TPU workloads run on the desired TPU machine type and topology.
For more details, see the [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/tpus#workload_preparation).

## Step 3: Configure `kubectl` to connect to the cluster

Run the following command to download Google Cloud credentials and configure the Kubernetes CLI to use them.

```sh
gcloud container clusters get-credentials kuberay-tpu-cluster --zone us-central2-b
```

For more details, see [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-access-for-kubectl).

## Step 4: Install the TPU initialization webhook

> GKE provides a mutating webhook to handle TPU pod scheduling and bootstrap certain environment variables used for [JAX](https://github.com/google/jax) initialization. The Ray TPU webhook is installed once per cluster and requires a Kuberay operator version of at least v1.1.0. For instructions on installing the webhook, see the documentation [here](https://github.com/GoogleCloudPlatform/ai-on-gke/tree/main/ray-on-gke/guides/tpu#manually-installing-the-tpu-initialization-webhook).

Reviewer: I think it'll be better to add the webhook installation steps in-line here instead of linking out to an external repo. Thoughts?

Author: Added in [4296124](https://github.com/ray-project/ray/pull/45924/commits/42961243c485652ae4d35de1aa3779f12b199aed).

GKE uses Kubernetes node selectors to ensure TPU workloads run on the desired machine type and topology.
For more details, see the [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/tpus#workload_preparation).

## Step 3: Configure `kubectl` to connect to the cluster

Run the following command to download Google Cloud credentials and configure the Kubernetes CLI to use them.

```sh
gcloud container clusters get-credentials kuberay-tpu-cluster --zone us-central2-b
```

For more details, see [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-access-for-kubectl).

## Step 4: Install the TPU initialization webhook

> GKE provides a [mutating webhook](https://github.com/GoogleCloudPlatform/ai-on-gke/tree/main/ray-on-gke/tpu/kuberay-tpu-webhook) to handle TPU pod scheduling and bootstrap certain environment variables used for [JAX](https://github.com/google/jax) initialization. The Ray TPU webhook is installed once per cluster and requires a Kuberay operator version of at least v1.1.0. The webhook requires [cert-manager](https://github.com/cert-manager/cert-manager) to be installed in-cluster to handle TLS certificate injection. cert-manager can be installed in both GKE standard and autopilot clusters using the following helm commands:

Author: The webhook can be used for both single-host and multi-host TPUs (although it is required for multi-host). We still set some environment variables for single-host TPUs such as `TPU_NAME` which is used [here](https://github.com/ray-project/ray/blob/8999fd5194c0f21f845d7bd059e39b6b7d680051/python/ray/util/accelerators/tpu.py#L12).

```
helm repo add jetstack https://charts.jetstack.io
helm repo update
helm install --create-namespace --namespace cert-manager --set installCRDs=true --set global.leaderElection.namespace=cert-manager cert-manager jetstack/cert-manager
```

> Installing the Ray TPU initialization webhook:

Reviewer 2: Is there any way to install the webhook without cloning the repo?

Author: When manually installing the webhook, it's currently necessary to clone the repo and run the make commands to install the deployments in `ai-on-gke/ray-on-gke/tpu/kuberay-tpu-webhook`.

Author: I created a PR to add a helm chart for the TPU webhook, which would remove the need to clone the repo. Once that's merged I can update the install instructions (either in this PR or in a follow-up PR).

(kuberay-gke-tpu-cluster-setup)=

# Start Google Cloud GKE Cluster with TPUs for KubeRay

See the [GKE documentation](<https://cloud.google.com/kubernetes-engine/docs/how-to/tpus>) for full details, or continue reading for a quick start.

## Step 1: Create a Kubernetes cluster on GKE

> Run all following commands on your local machine or on the [Google Cloud Shell](https://cloud.google.com/shell). If running from your local machine, you need to install the [Google Cloud SDK](https://cloud.google.com/sdk/docs/install).

Reviewer 3: ```suggestion
Run the following commands on your local machine or on the [Google Cloud Shell](https://cloud.google.com/shell). If running from your local machine, install the [Google Cloud SDK](https://cloud.google.com/sdk/docs/install).
```

> Create a Standard GKE cluster with the Ray Operator enabled in the us-central2-b compute region:

Reviewer 3: ```suggestion
Create a Standard GKE cluster and enable the Ray Operator in the us-central2-b compute region:
```

  --tpu-topology 2x2x2
```

The `--tpu-topology` flag specifies the physical topology of the TPU Pod slice. This example uses a v4 TPU slice with either a 2x2x1 or 2x2x2 topology. v4 TPUs have 4 chips per VM host, so a 2x2x2 v4 slice has 8 chips total and 2 TPU hosts, each scheduled on their own node. GKE treats multi-host TPU slices as atomic units, and scales them using node pools rather than singular nodes. Therefore, the number of TPU hosts should always equal the number of nodes in the TPU node pool. For more information about selecting a TPU topology and accelerator, see the [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/concepts/tpus).

GKE uses Kubernetes node selectors to ensure TPU workloads run on the desired machine type and topology.
For more details, see the [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/tpus#workload_preparation).

## Step 2: Configure `kubectl` to connect to the cluster

Run the following command to download Google Cloud credentials and configure the Kubernetes CLI to use them.

```sh
gcloud container clusters get-credentials kuberay-tpu-cluster --zone us-central2-b
```

> For more details, see [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-access-for-kubectl).

Reviewer 3: ```suggestion
For more details, see the [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-access-for-kubectl).
```

## Step 3: Install the TPU initialization webhook

GKE provides a [validating and mutating webhook](https://github.com/GoogleCloudPlatform/ai-on-gke/tree/main/ray-on-gke/tpu/kuberay-tpu-webhook) to handle TPU Pod scheduling and bootstrap certain environment variables used for [JAX](https://github.com/google/jax) initialization. The Ray TPU webhook requires a KubeRay operator version of at least v1.1.0. GKE automatically installs the Ray TPU webhook through the [Ray Operator Addon](https://cloud.google.com/kubernetes-engine/docs/add-on/ray-on-gke/how-to/enable-ray-on-gke) with GKE versions 1.30.0-gke.1747000 or later. When manually installing the Ray TPU webhook, you need [cert-manager](https://github.com/cert-manager/cert-manager) to handle TLS certificate injection. You can install cert-manager in both GKE Standard and Autopilot clusters using the following helm commands:

> ### [Optional] Manual install in a GKE cluster without the Ray Addon:

Reviewer 3: ```suggestion
### [Optional] Manually install in a GKE cluster without the Ray Addon:
```

(kuberay-gke-tpu-cluster-setup)=

# Start Google Cloud GKE Cluster with TPUs for KubeRay

See the [GKE documentation](<https://cloud.google.com/kubernetes-engine/docs/how-to/tpus>) for full details, or continue reading for a quick start.

## Step 1: Create a Kubernetes cluster on GKE

Run all following commands on your local machine or on the [Google Cloud Shell](https://cloud.google.com/shell). If running from your local machine, you need to install the [Google Cloud SDK](https://cloud.google.com/sdk/docs/install).

### GKE Autopilot

Create an Autopilot GKE cluster with the [Ray Operator](https://cloud.google.com/kubernetes-engine/docs/add-on/ray-on-gke/concepts/overview) enabled in the us-central2-b compute region:

```sh
gcloud container clusters create-auto kuberay-tpu-cluster \
    --enable-ray-operator \

>     --cluster-version=CLUSTER_VERSION \

Reviewer: since cluster version is 1.30, for now we need to specify ` --release-channel rapid`

Author: This flag does not appear to be needed for the Standard cluster CLI command (ie. `gcloud container clusters create ...`), I removed the Autopilot commands in [d1c3338](https://github.com/ray-project/ray/pull/45924/commits/d1c33386da6d1bfa18dfa76ec98fa157af93d05e) based on the offline discussion about ephemeral storage limits, but will add these instructions back in with the `--release-channel rapid` flag once that's resolved.

    --location=us-central2
```

> Replace CLUSTER_VERSION with the GKE version for the cluster, which must be version 1.30.1 or later. Alternatively, create a cluster using the [Google Cloud Console](https://console.cloud.google.com/kubernetes/list).

Reviewer 2: nit
```suggestion
Replace `CLUSTER_VERSION` with the GKE version for the cluster, which must be version 1.30.1 or later. Alternatively, create a cluster using the [Google Cloud Console](https://console.cloud.google.com/kubernetes/list).
```

To use TPUs with Autopilot mode, you must select a compute location with capacity for TPU accelerators. v4 TPUs are available in us-central2-b. For more information on TPU availability by region/zone, see the [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/concepts/tpus#availability).

### GKE Standard

Create a Standard GKE cluster with the Ray Operator enabled in the us-central2-b compute region:

```sh
gcloud container clusters create kuberay-tpu-cluster \
    --addons=RayOperator \

>     --cluster-version=VERSION \

Author: Done in d1c3338

GKE uses Kubernetes node selectors to ensure TPU workloads run on the desired machine type and topology.
For more details, see the [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/tpus#workload_preparation).

## Step 2: Configure `kubectl` to connect to the cluster

Run the following command to download Google Cloud credentials and configure the Kubernetes CLI to use them.

```sh
gcloud container clusters get-credentials kuberay-tpu-cluster --zone us-central2-b
```

For more details, see [GKE documentation](https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-access-for-kubectl).

## Step 3: Install the TPU initialization webhook

> GKE provides a [validating and mutating webhook](https://github.com/GoogleCloudPlatform/ai-on-gke/tree/main/ray-on-gke/tpu/kuberay-tpu-webhook) to handle TPU Pod scheduling and bootstrap certain environment variables used for [JAX](https://github.com/google/jax) initialization. The Ray TPU webhook is installed once per cluster and requires a Kuberay operator version of at least v1.1.0. The Ray TPU webhook is automatically installed through the [Ray Operator Addon](https://cloud.google.com/kubernetes-engine/docs/add-on/ray-on-gke/how-to/enable-ray-on-gke) with GKE versions 1.30.0-gke.1747000 or later. When manually installing the Ray TPU webhook, [cert-manager](https://github.com/cert-manager/cert-manager) is required to handle TLS certificate injection. cert-manager can be installed in both GKE Standard and Autopilot clusters using the following helm commands:

Author: GKE installs the Ray TPU webhook not KubeRay, I'll add these changes but edit KubeRay to GKE in another commit where appropriate.
